gremlin/ui/device.py

```diff
@@ -30,13 +32,164 @@
+class DeviceDatabase:
+    """Provides button/axis/hat to name mapping for known devices"""
+
+    def __init__(self) -> None:
+        self._load()
+
+    def _load(self) -> None:
+
+        db_file = util.resource_path("device_db.json")
+        if not util.file_exists_and_is_accessible(db_file):
+            return
+
+        load_successful = False
+        json_data = {}
+        try:
+            json_data = json.load(open(db_file))
+            load_successful = True
+        except JSONDecodeError as error:
+            logging.getLogger("system").error(
+                    f"There was an error loading {db_file}: {error}")
+
+        parser_successful = self._parse_device_db(json_data)
+
+        if load_successful and parser_successful:
+            self._device_db = json_data
+        else:
+            self._device_db = {"revision": 0, "devices": [], "mapping": {}}
+
+    def _parse_device_db(self, device_db: Dict[str, Any]) -> bool:
+        """
+        Parse device_db to make sure it has a valid structure.
+
+        This piece could be refactored with JSON schema validation code,
+        if need be in the future.
+        """
+        parser_successful = True
+
+        # check top structure: devices and mappings
+        if (
+            device_db.get("mapping", None) is None or
+            device_db.get("devices", None) is None or
+            not isinstance(device_db.get("mapping", []), dict) or
+            not isinstance(device_db.get("devices", []), list)
+        ):
+            logging.getLogger("system").error(
+                    "DeviceDatabase is corrupt and/or is missing mapping or device information.")
+            parser_successful = False
+
+        # check devices contain mandatory fields
+        device_count = 0
+        for dev in device_db.get("devices", []):
+            if (
+                dev.get("product_id", None) is None or
+                dev.get("vendor_id", None) is None or
+                dev.get("mapping", None) is None
+            ):
+                logging.getLogger("system").error(
+                    f"DeviceDatabase device structure is corrupt for device {dev}")
+                parser_successful = False
+            else:
+                device_count += 1
+
+        mapping_count = 0
+        for mapping in device_db.get("mapping", []):
+            mapping_count += 1
+
+        if parser_successful and device_count > 0 and mapping_count > 0:
+            return True
+
+        return False
+
+    def get_mapping(self, device: dill.DeviceSummary) -> DeviceMapping | None:
+        """Returns: the button/axis <-> name mapping of a device"""
+        if device is None:
+            return None
+
+        for dev in self._device_db.get("devices", []):
+            if (
+                dev.get("product_id", None) == device.product_id and
+                dev.get("vendor_id", None) == device.vendor_id
+            ):
+                if (
+                    "mapping" not in dev or
```

Comment:
`"mapping" not in dev` should not occur anymore due to the `_parse_device_db` check, correct?